### ISSUE TITLE
Randomly failing travis test

### ISSUE THREAD

**Issue by indigoxela** [backdrop-contrib/b#89]
The test for b fails randomly.
One commit passes, the next one doesn't ...

Error message:

```
Failed asserting that 'file_get_contents(https://api.github.com/repos/backdrop-contrib/simplify): failed to open stream: HTTP request failed! HTTP/1.1 403 Forbidden\r\n

/home/travis/build/backdrop-contrib/b/commands/PM/download.b.inc:224\n                                                                                                 
```
No clue, how this should get fixed.

**Comment by ghost** [comment]
Because the part that's failing is getting repo info from GitHub, and because it sometimes works and sometimes dosnt, I suspect its to do with rate limiting or just GitHub having issues. I don't see how there's anything we can do to fix this in `b` (correct me if I'm wrong).

**Comment by indigoxela** [comment]
> I suspect its to do with rate limiting or just GitHub having issues

That's what I also suspected. Github status didn't give me any clue. https://www.githubstatus.com/

Hm... Let's close this issue. We can't solve it anyway.

**Comment by ghost** [comment]
I'm getting this too now, so maybe we're hitting a limit with downloading from GitHub or something... Need to look into it.

**Comment by ghost** [comment]
I wonder what the benefit of using the GitHub API is... I mean, we could just download projects from: `https://github.com/backdrop-contrib/[NAME]/releases/latest/download/[NAME].zip`

E.g.: https://github.com/backdrop-contrib/simplify/releases/latest/download/simplify.zip

@Gormartsen I notice that you wrote the original version of the Download command. Can you weigh in on why the GitHub API was chosen, and if you see any issues with using a direct `.zip` URL instead?

Basically trying to avoid the API's rate limit (which I assume is the issue here) and also avoid users needing to provide their own GitHub API tokens, etc. for authorisation (which is the other way to avoid the limit).

**Comment by Gormartsen** [comment]
@BWPanda i checked code. So yes. Part of it was validation, another part is that we can guarantee correct link to download until API changes. Not a big deal. You can replace by sending your way request and if not 200 response, then not found.

Also there is failback to download via git clone I assume?

It was long time ago. I would check a download limit by URL before go with changes.

**Comment by ghost** [comment]
Yes, I'd leave the git clone fallback in place. I looked for URL download limits, but couldn't find anything about that. So I guess I'll just try using it instead and then see if this issue persists...

**Comment by ghost** [comment]
Here's a PR that replaces GitHub API URLs with a simple `get_headers()` call to check if a URL exists before attempting to download from it: https://github.com/backdrop-contrib/b/pull/91

Testing welcome!

**Comment by indigoxela** [comment]
Looks good. Of course, because of the nature of this issue (_random_ failures), it's hard to verify reliably.

Downloading works. :wink:
